api/src/lib/records.js

```diff
@@ -0,0 +1,298 @@
+    logger.info(
+      `recordsForUpload(${upload.id}) initializing req.recordsForUpload cache`
+    )
+    req.recordsForUpload = {}
+  }
+
+  if (req.recordsForUpload[upload.id]) {
+    logger.info(`recordsForUpload(${upload.id}): reading from cache`)
+    return req.recordsForUpload[upload.id]
+  }
+
+  logger.info(`recordsForUpload(${upload.id}): reading from disk`)
+  const recordPromise = loadRecordsForUpload(upload)
+
+  // By caching the promise, we ensure that parallel fetches won't start a new
+  // filesystem read, even if the first read hasn't resolved yet.
+  req.recordsForUpload[upload.id] = recordPromise
+
+  return recordPromise
+}
+
+// async function recordsForReportingPeriod(periodId, tenantId) {
+//   logger.info(`recordsForReportingPeriod(${periodId})`)
+//   requiredArgument(
+//     periodId,
+//     'must specify periodId in recordsForReportingPeriod'
+//   )
+
+//   const uploads = await usedForTreasuryExport(periodId, tenantId)
+//   const groupedRecords = await Promise.all(
+//     uploads.map((upload) => recordsForUpload(upload))
+//   )
+//   return groupedRecords.flat()
+// }
+
+/**
+ * Get the most recent, validated record for each unique project, as of the
+ * specified reporting period.
+ */
+// async function mostRecentProjectRecords(periodId, tenantId) {
+//   logger.info(`mostRecentProjectRecords(${periodId})`)
+//   requiredArgument(
+//     periodId,
+//     'must specify periodId in mostRecentProjectRecords'
+//   )
+
+//   const reportingPeriods = await previousReportingPeriods(
+//     periodId,
+//     organizationId
+//   )
+
+//   const allRecords = await Promise.all(
+//     reportingPeriods.map(({ id }) => recordsForReportingPeriod(id, tenantId))
+//   )
+
+//   const latestProjectRecords = allRecords
+//     .flat()
+//     // exclude non-project records
+//     .filter((record) => Object.values(EC_SHEET_TYPES).includes(record.type))
+//     // collect the latest record for each project ID
+//     .reduce((accumulator, record) => {
+//       accumulator[record.content.Project_Identification_Number__c] = record
+//       return accumulator
+//     }, {})
+
+//   return Object.values(latestProjectRecords)
+// }
+
+// async function recordsForProject(periodId, tenantId) {
+//   log(`recordsForProject`)
+//   requiredArgument(
+//     periodId,
+//     'must specify periodId in mostRecentProjectRecords'
+//   )
+
+//   const reportingPeriods = await getPreviousReportingPeriods(
+//     periodId,
+//     undefined,
+//     tenantId
+//   )
+
+//   const allRecords = await Promise.all(
+//     reportingPeriods.map(({ id }) => recordsForReportingPeriod(id, tenantId))
+//   )
+
+//   const projectRecords = allRecords
+//     .flat()
+//     // exclude non-project records
+//     .filter((record) =>
+//       [
+//         ...Object.values(EC_SHEET_TYPES),
+//         'awards50k',
+//         'expenditures50k',
+//       ].includes(record.type)
+//     )
+
+//   return Object.values(projectRecords)
+// }
+
```

Comment:
I commented out a fair amount of code because it isn't needed immediately for validating uploads.